### DISCUSSION TITLE
Firmware update

### DISCUSSION THREAD

**Discussion by nobodyguy** [semuconsulting/PyGPSClient#58] [Q&A] [answered]
Hi, thank you for this awesome project!
Have anyone tried to dig deeper into a possibility to reverse engineer a firmware update mechanism?

**Comment by semuadmin** [comment, marked as answer]
Hi @nobodyguy 

So, yes, I have looked into this as I'm conscious it's one of the few things you can do in u-center that you can't currently do in PyGPSClient. 

I have managed to perform a home-grown firmware upgrade on an old M7 device, based on info gleaned from the u-blox technical support forum.

While it's technically feasible, there are a number of significant challenges: 

1. Establishing which particular MCU is used in each u-blox device, and keeping track of any significant hardware changes that u-blox might implement along the way. This information is not generally in the public domain.
2. Maintaining an appropriate deployment toolchain for each device (u-blox use a variety of MCUs).
3. It would probably involve calling out to one or more C or C++ libraries, which adds significant complexity to the Python code.
4. CRITICAL - Testing it to a level where we can be confident that we're not going to be bricking users' devices. This would entail access to every device and/or firmware variant in u-blox's GNSS product catalogue (_with spares to replace any bricked in testing!_), which would be prohibitively expensive for an unpaid volunteer project like PyGPSClient.
5. There may be open-source license implications - the UBX protocol is effectively in the public domain, but the firmware specifications are not and access may be subject to NDA.

So as much as I'd like to add this functionality, in practice the PyGPSClient project simply doesn't have the time or resources to do it.

Sorry to disappoint.

**Reply by adamgarbo** [reply]
I came to Discussions specifically for this answer. Disappointing, but completely understandable. Frustrating that u-blox is so short-sighted they willingly choose not to support macOS/Linux users.

**Reply by nobodyguy** [reply]
Forgot to reply sooner, sorry. Yes, that's understandable, thank you 🙂 I guess ucenter inside Docker container with Wine will have to do.

**Comment by semuadmin** [comment]
Hi @adamgarbo @nobodyguy 

See also discussion https://github.com/semuconsulting/pyubx2/discussions/120, which includes u-blox's response stating that the inclusion of the relevant UBX firmware update message definitions would violate the terms of use.

I'll keep it on the Backlog in case this position ever changes.
